usl_pipeline/cloud_functions/main.py

```diff
@@ -675,24 +675,136 @@ def _compute_custom_wps_variables(dataset: xarray.Dataset) -> xarray.Dataset:
+        # Extract hours and minutes from the datetime objects
+        utc_hours_minutes = numpy.array([t.astype('datetime64[h]').astype(int) % 24 + t.astype('datetime64[m]').astype(int) % 60 / 60 for t in times])
+
+        # Define the function to calculate solar time
+        def calculate_solar_time(utc_time, longitude):
+            return (utc_time + longitude / 15) % 24
+
+        # Use the correct variable names for latitude and longitude coordinates
+        latitude_coord = 'XLAT_M'
+        longitude_coord = 'XLONG_M'
+
+        # Extract latitude and longitude coordinates
+        latitudes = dataset[latitude_coord][0, :, :]
+        longitudes = dataset[longitude_coord][0, :, :]
+
+        # Create an empty DataArray to store solar times
+        solar_times = xarray.DataArray(
+            numpy.zeros((len(utc_hours_minutes),) + longitude.shape[1:]),
+            # Shape should match (utc_time, south_north, west_east)
+            coords={
+                'utc_time': utc_hours_minutes,
+                'south_north': latitudes.coords['south_north'],
+                'west_east': longitudes.coords['west_east']
+            },
+            dims=['utc_time', 'south_north', 'west_east']
+        )
+
+        # Calculate solar time for each extracted UTC time
+        for i, utc_time in enumerate(utc_hours_minutes):
+            solar_times.loc[utc_time, :, :] = calculate_solar_time(utc_time, longitudes)
+
+        # Define the function to convert time in hours to sine value
+        def time_to_sine(hours):
+            # Map the hours (in 24-hour format) to radians [0, 2π]
+            radians = (hours / 24.0) * 2 * numpy.pi
+            return numpy.sin(radians)
+
+        # Define the function to convert time in hours to cosine value
+        def time_to_cosine(hours):
+            # Map the hours (in 24-hour format) to radians [0, 2π]
+            radians = (hours / 24.0) * 2 * numpy.pi
+            return numpy.cos(radians)
+
+        # Create empty DataArrays to store sine and cosine values of solar times
+        sin_solar_time = xarray.DataArray(
+            numpy.zeros_like(solar_times),
+            coords=solar_times.coords,
+            dims=solar_times.dims
+        )
+
+        cos_solar_time = xarray.DataArray(
+            numpy.zeros_like(solar_times),
+            coords=solar_times.coords,
+            dims=solar_times.dims
+        )
+
+        # Convert solar times to sine and cosine values
+        for utc_time in utc_hours_minutes:
+            sin_solar_time.loc[utc_time, :, :] = time_to_sine(solar_times.sel(utc_time=utc_time))
+            cos_solar_time.loc[utc_time, :, :] = time_to_cosine(solar_times.sel(utc_time=utc_time))
+            
+        new_dims = ["Time", "south_north", "west_east"]
+        dataset = dataset.assign(SOLAR_TIME_SIN=(new_dims, sin_solar_time))
+        dataset = dataset.assign(SOLAR_TIME_COS=(new_dims, cos_solar_time))
```

Comment:
Is that the same as what we did yesterday? should i re-check or it is the same?

Reply:
I added your line of code of fetching time from the file. The rest is my previous code. But I had to make it align with the way the rest of the code was written in main.py. Especially concerned if the way if I have called dataset functions is correct or not @saskatchuwan or @waltaskew pls have a look